File: crates/macos_window_manager/swift/Sources/MacosWindowManagerFFI/Transport.swift
import MacosWindowManagerCore

public let MWM_STATUS_OK: Int32 = 0
public let MWM_STATUS_INVALID_ARGUMENT: Int32 = 1
public let MWM_STATUS_UNAVAILABLE: Int32 = 2
public let MWM_STATUS_CONNECT_MISSING_REQUIRED_SYMBOL: Int32 = 10
public let MWM_STATUS_CONNECT_MISSING_ACCESSIBILITY_PERMISSION: Int32 = 11
public let MWM_STATUS_CONNECT_MISSING_TOPOLOGY_PRECONDITION: Int32 = 12
public let MWM_STATUS_PROBE_MISSING_TOPOLOGY: Int32 = 20

/// Runtime ABI guards for the transport shared with Rust `src/transport.rs`.
///
/// Pointer payloads are owned by the Swift FFI layer when non-nil and must be
/// released by Rust via `mwm_status_release` or `mwm_desktop_snapshot_release`
/// after copying any needed values out of the transport structs.
@inline(__always)
func verifyTransportAbiContract() {
    precondition(MemoryLayout<MwmStatus>.stride == 16)
    precondition(MemoryLayout<MwmStatus>.alignment == 8)
    precondition(MemoryLayout<MwmStatus>.offset(of: \.code) == 0)
    precondition(MemoryLayout<MwmStatus>.offset(of: \.message_ptr) == 8)

    precondition(MemoryLayout<MwmRectAbi>.stride == 16)
    precondition(MemoryLayout<MwmRectAbi>.alignment == 4)
    precondition(MemoryLayout<MwmRectAbi>.offset(of: \.x) == 0)
    precondition(MemoryLayout<MwmRectAbi>.offset(of: \.y) == 4)
    precondition(MemoryLayout<MwmRectAbi>.offset(of: \.width) == 8)
    precondition(MemoryLayout<MwmRectAbi>.offset(of: \.height) == 12)

    precondition(MemoryLayout<MwmSpaceAbi>.stride == 24)
    precondition(MemoryLayout<MwmSpaceAbi>.alignment == 8)
    precondition(MemoryLayout<MwmSpaceAbi>.offset(of: \.id) == 0)
    precondition(MemoryLayout<MwmSpaceAbi>.offset(of: \.display_index) == 8)
    precondition(MemoryLayout<MwmSpaceAbi>.offset(of: \.active) == 16)
    precondition(MemoryLayout<MwmSpaceAbi>.offset(of: \.kind) == 20)

    precondition(MemoryLayout<MwmWindowAbi>.stride == 80)
    precondition(MemoryLayout<MwmWindowAbi>.alignment == 8)
    precondition(MemoryLayout<MwmWindowAbi>.offset(of: \.id) == 0)
    precondition(MemoryLayout<MwmWindowAbi>.offset(of: \.pid) == 8)
    precondition(MemoryLayout<MwmWindowAbi>.offset(of: \.has_pid) == 12)
    precondition(MemoryLayout<MwmWindowAbi>.offset(of: \.app_id_ptr) == 16)
    precondition(MemoryLayout<MwmWindowAbi>.offset(of: \.title_ptr) == 24)
    precondition(MemoryLayout<MwmWindowAbi>.offset(of: \.frame) == 32)
    precondition(MemoryLayout<MwmWindowAbi>.offset(of: \.has_frame) == 48)
    precondition(MemoryLayout<MwmWindowAbi>.offset(of: \.level) == 52)
    precondition(MemoryLayout<MwmWindowAbi>.offset(of: \.space_id) == 56)
    precondition(MemoryLayout<MwmWindowAbi>.offset(of: \.order_index) == 64)
    precondition(MemoryLayout<MwmWindowAbi>.offset(of: \.has_order_index) == 72)

    precondition(MemoryLayout<MwmDesktopSnapshotAbi>.stride == 40)
    precondition(MemoryLayout<MwmDesktopSnapshotAbi>.alignment == 8)
    precondition(MemoryLayout<MwmDesktopSnapshotAbi>.offset(of: \.spaces_ptr) == 0)
    precondition(MemoryLayout<MwmDesktopSnapshotAbi>.offset(of: \.spaces_len) == 8)
    precondition(MemoryLayout<MwmDesktopSnapshotAbi>.offset(of: \.windows_ptr) == 16)
    precondition(MemoryLayout<MwmDesktopSnapshotAbi>.offset(of: \.windows_len) == 24)
    precondition(MemoryLayout<MwmDesktopSnapshotAbi>.offset(of: \.focused_window_id) == 32)
}

/// FFI status transport shared with Rust.
///
/// `message_ptr` is owned by the Swift FFI layer when non-nil. Rust must copy
/// the string and then call `mwm_status_release` to release the owned payload.
public struct MwmStatus {
    public var code: Int32
    public var message_ptr: UnsafeMutablePointer<CChar>?

    public init(
        code: Int32 = MWM_STATUS_OK,
        message_ptr: UnsafeMutablePointer<CChar>? = nil
    ) {
        self.code = code
        self.message_ptr = message_ptr
    }
}

/// FFI rectangle transport shared with Rust.
public struct MwmRectAbi {
    public var x: Int32
    public var y: Int32
    public var width: Int32
    public var height: Int32

    public init(x: Int32 = 0, y: Int32 = 0, width: Int32 = 0, height: Int32 = 0) {
        self.x = x
        self.y = y
        self.width = width
        self.height = height
    }
}

/// FFI space transport shared with Rust.
public struct MwmSpaceAbi {
    public var id: UInt64
    public var display_index: Int
    public var active: UInt8
    public var kind: Int32

    public init(
        id: UInt64 = 0,
        display_index: Int = 0,
        active: UInt8 = 0,
        kind: Int32 = 0
    ) {
        self.id = id
        self.display_index = display_index
        self.active = active
        self.kind = kind
    }
}

/// FFI window transport shared with Rust.
///
/// `app_id_ptr` and `title_ptr` are owned by the Swift FFI layer when non-nil
/// and are released as part of `mwm_desktop_snapshot_release`.
public struct MwmWindowAbi {
    public var id: UInt64
    public var pid: UInt32
    public var has_pid: UInt8
    public var app_id_ptr: UnsafeMutablePointer<CChar>?
    public var title_ptr: UnsafeMutablePointer<CChar>?
    public var frame: MwmRectAbi
    public var has_frame: UInt8
    public var level: Int32
    public var space_id: UInt64
    public var order_index: Int
    public var has_order_index: UInt8

    public init(
        id: UInt64 = 0,
        pid: UInt32 = 0,
        has_pid: UInt8 = 0,
        app_id_ptr: UnsafeMutablePointer<CChar>? = nil,
        title_ptr: UnsafeMutablePointer<CChar>? = nil,
        frame: MwmRectAbi = MwmRectAbi(),
        has_frame: UInt8 = 0,
        level: Int32 = 0,
        space_id: UInt64 = 0,
        order_index: Int = 0,
        has_order_index: UInt8 = 0
    ) {
        self.id = id
        self.pid = pid
        self.has_pid = has_pid
        self.app_id_ptr = app_id_ptr
        self.title_ptr = title_ptr
        self.frame = frame
        self.has_frame = has_frame
        self.level = level
        self.space_id = space_id
        self.order_index = order_index
        self.has_order_index = has_order_index
    }
}

/// FFI desktop snapshot transport shared with Rust.
///
/// Any non-nil pointer fields are owned by the Swift FFI layer and must be
/// released by Rust via `mwm_desktop_snapshot_release` after copying the
/// snapshot contents into Rust-owned structures.
public struct MwmDesktopSnapshotAbi {
    public var spaces_ptr: UnsafeMutablePointer<MwmSpaceAbi>?
    public var spaces_len: Int
    public var windows_ptr: UnsafeMutablePointer<MwmWindowAbi>?
    public var windows_len: Int
    public var focused_window_id: UInt64

    public init(
        spaces_ptr: UnsafeMutablePointer<MwmSpaceAbi>? = nil,
        spaces_len: Int = 0,
        windows_ptr: UnsafeMutablePointer<MwmWindowAbi>? = nil,
        windows_len: Int = 0,
        focused_window_id: UInt64 = 0
    ) {
        self.spaces_ptr = spaces_ptr
        self.spaces_len = spaces_len
        self.windows_ptr = windows_ptr
        self.windows_len = windows_len
        self.focused_window_id = focused_window_id
    }
}

extension MwmStatus {
    mutating func releaseOwnedPayloads() {
        message_ptr?.deallocate()
        self = MwmStatus()
    }
}

extension MwmWindowAbi {
    mutating func releaseOwnedPayloads() {
        app_id_ptr?.deallocate()
        title_ptr?.deallocate()
        self = MwmWindowAbi(
            id: id,
            pid: pid,
            has_pid: has_pid,
            app_id_ptr: nil,
            title_ptr: nil,
            frame: frame,
            has_frame: has_frame,
            level: level,
            space_id: space_id,
            order_index: order_index,
            has_order_index: has_order_index
        )
    }
}

extension MwmDesktopSnapshotAbi {
    mutating func releaseOwnedPayloads() {
        if let windows_ptr {
            for index in 0..<windows_len {
                windows_ptr.advanced(by: index).pointee.releaseOwnedPayloads()
            }
            windows_ptr.deallocate()
        }

        spaces_ptr?.deallocate()
        self = MwmDesktopSnapshotAbi()
    }
}

extension MwmRectAbi {
    init(_ bounds: NativeBounds) {
        self.init(
            x: bounds.x,
            y: bounds.y,
            width: bounds.width,
            height: bounds.height
        )
    }
}

extension MwmSpaceAbi {
    init(_ space: DesktopSpaceSnapshot) {
        self.init(
            id: space.id,
            display_index: space.displayIndex,
            active: space.active ? 1 : 0,
            kind: space.kind.rawValue
        )
    }
}

extension MwmWindowAbi {
    init(_ window: DesktopWindowSnapshot) {
        self.init(
            id: window.id,
            pid: window.pid ?? 0,
            has_pid: window.pid == nil ? 0 : 1,
            app_id_ptr: window.appID?.ownedCString(),
            title_ptr: window.title?.ownedCString(),
            frame: window.bounds.map(MwmRectAbi.init) ?? MwmRectAbi(),
            has_frame: window.bounds == nil ? 0 : 1,
            level: window.level,
            space_id: window.spaceID,
            order_index: window.orderIndex ?? 0,
            has_order_index: window.orderIndex == nil ? 0 : 1
        )
    }
}

extension MwmDesktopSnapshotAbi {
    init(_ snapshot: DesktopSnapshot) {
        let spacesPointer: UnsafeMutablePointer<MwmSpaceAbi>?
        if snapshot.spaces.isEmpty {
            spacesPointer = nil
        } else {
            let pointer = UnsafeMutablePointer<MwmSpaceAbi>.allocate(capacity: snapshot.spaces.count)
            pointer.initialize(from: snapshot.spaces.map(MwmSpaceAbi.init), count: snapshot.spaces.count)
            spacesPointer = pointer
        }

        let windowsPointer: UnsafeMutablePointer<MwmWindowAbi>?
        if snapshot.windows.isEmpty {
            windowsPointer = nil
        } else {
            let pointer = UnsafeMutablePointer<MwmWindowAbi>.allocate(capacity: snapshot.windows.count)
            pointer.initialize(from: snapshot.windows.map(MwmWindowAbi.init), count: snapshot.windows.count)
            windowsPointer = pointer
        }

        self.init(
            spaces_ptr: spacesPointer,
            spaces_len: snapshot.spaces.count,
            windows_ptr: windowsPointer,
            windows_len: snapshot.windows.count,
            focused_window_id: snapshot.focusedWindowID ?? 0
        )
    }
}

extension String {
    func ownedCString() -> UnsafeMutablePointer<CChar> {
        let buffer = UnsafeMutablePointer<CChar>.allocate(capacity: utf8.count + 1)
        utf8CString.withUnsafeBufferPointer { value in
            buffer.initialize(from: value.baseAddress!, count: value.count)
        }
        return buffer
    }
}

File: crates/macos_window_manager/src/transport.rs
#![allow(dead_code)]

use std::ffi::{c_char, CStr};

use crate::{
    desktop_topology_snapshot::SpaceKind, NativeBounds, NativeDesktopSnapshot, NativeSpaceSnapshot,
    NativeWindowSnapshot,
};

pub(crate) const MWM_STATUS_OK: i32 = 0;
pub(crate) const MWM_STATUS_INVALID_ARGUMENT: i32 = 1;
pub(crate) const MWM_STATUS_UNAVAILABLE: i32 = 2;
pub(crate) const MWM_STATUS_CONNECT_MISSING_REQUIRED_SYMBOL: i32 = 10;
pub(crate) const MWM_STATUS_CONNECT_MISSING_ACCESSIBILITY_PERMISSION: i32 = 11;
pub(crate) const MWM_STATUS_CONNECT_MISSING_TOPOLOGY_PRECONDITION: i32 = 12;
pub(crate) const MWM_STATUS_PROBE_MISSING_TOPOLOGY: i32 = 20;

/// ABI layout assertions for the Swift transport contract.
///
/// The Swift definitions in `swift/Sources/MacosWindowManagerFFI/Transport.swift`
/// must keep these exact sizes, alignments, and field offsets on macOS 64-bit.
/// Pointer payloads are owned by the Swift FFI layer when non-null and must be
/// released via `mwm_status_release` or `mwm_desktop_snapshot_release` after the
/// Rust side has copied any needed data out of them.
#[cfg(not(target_pointer_width = "64"))]
compile_error!("macos_window_manager FFI transport requires a 64-bit target");
#[cfg(target_pointer_width = "64")]
const _: [(); 16] = [(); std::mem::size_of::<MwmStatus>()];
#[cfg(target_pointer_width = "64")]
const _: [(); 8] = [(); std::mem::align_of::<MwmStatus>()];
#[cfg(target_pointer_width = "64")]
const _: [(); 0] = [(); std::mem::offset_of!(MwmStatus, code)];
#[cfg(target_pointer_width = "64")]
const _: [(); 8] = [(); std::mem::offset_of!(MwmStatus, message_ptr)];
#[cfg(target_pointer_width = "64")]
const _: [(); 16] = [(); std::mem::size_of::<MwmRectAbi>()];
#[cfg(target_pointer_width = "64")]
const _: [(); 4] = [(); std::mem::align_of::<MwmRectAbi>()];
#[cfg(target_pointer_width = "64")]
const _: [(); 0] = [(); std::mem::offset_of!(MwmRectAbi, x)];
#[cfg(target_pointer_width = "64")]
const _: [(); 4] = [(); std::mem::offset_of!(MwmRectAbi, y)];
#[cfg(target_pointer_width = "64")]
const _: [(); 8] = [(); std::mem::offset_of!(MwmRectAbi, width)];
#[cfg(target_pointer_width = "64")]
const _: [(); 12] = [(); std::mem::offset_of!(MwmRectAbi, height)];
#[cfg(target_pointer_width = "64")]
const _: [(); 24] = [(); std::mem::size_of::<MwmSpaceAbi>()];
#[cfg(target_pointer_width = "64")]
const _: [(); 8] = [(); std::mem::align_of::<MwmSpaceAbi>()];
#[cfg(target_pointer_width = "64")]
const _: [(); 0] = [(); std::mem::offset_of!(MwmSpaceAbi, id)];
#[cfg(target_pointer_width = "64")]
const _: [(); 8] = [(); std::mem::offset_of!(MwmSpaceAbi, display_index)];
#[cfg(target_pointer_width = "64")]
const _: [(); 16] = [(); std::mem::offset_of!(MwmSpaceAbi, active)];
#[cfg(target_pointer_width = "64")]
const _: [(); 20] = [(); std::mem::offset_of!(MwmSpaceAbi, kind)];
#[cfg(target_pointer_width = "64")]
const _: [(); 80] = [(); std::mem::size_of::<MwmWindowAbi>()];
#[cfg(target_pointer_width = "64")]
const _: [(); 8] = [(); std::mem::align_of::<MwmWindowAbi>()];
#[cfg(target_pointer_width = "64")]
const _: [(); 0] = [(); std::mem::offset_of!(MwmWindowAbi, id)];
#[cfg(target_pointer_width = "64")]
const _: [(); 8] = [(); std::mem::offset_of!(MwmWindowAbi, pid)];
#[cfg(target_pointer_width = "64")]
const _: [(); 12] = [(); std::mem::offset_of!(MwmWindowAbi, has_pid)];
#[cfg(target_pointer_width = "64")]
const _: [(); 16] = [(); std::mem::offset_of!(MwmWindowAbi, app_id_ptr)];
#[cfg(target_pointer_width = "64")]
const _: [(); 24] = [(); std::mem::offset_of!(MwmWindowAbi, title_ptr)];
#[cfg(target_pointer_width = "64")]
const _: [(); 32] = [(); std::mem::offset_of!(MwmWindowAbi, frame)];
#[cfg(target_pointer_width = "64")]
const _: [(); 48] = [(); std::mem::offset_of!(MwmWindowAbi, has_frame)];
#[cfg(target_pointer_width = "64")]
const _: [(); 52] = [(); std::mem::offset_of!(MwmWindowAbi, level)];
#[cfg(target_pointer_width = "64")]
const _: [(); 56] = [(); std::mem::offset_of!(MwmWindowAbi, space_id)];
#[cfg(target_pointer_width = "64")]
const _: [(); 64] = [(); std::mem::offset_of!(MwmWindowAbi, order_index)];
#[cfg(target_pointer_width = "64")]
const _: [(); 72] = [(); std::mem::offset_of!(MwmWindowAbi, has_order_index)];
#[cfg(target_pointer_width = "64")]
const _: [(); 40] = [(); std::mem::size_of::<MwmDesktopSnapshotAbi>()];
#[cfg(target_pointer_width = "64")]
const _: [(); 8] = [(); std::mem::align_of::<MwmDesktopSnapshotAbi>()];
#[cfg(target_pointer_width = "64")]
const _: [(); 0] = [(); std::mem::offset_of!(MwmDesktopSnapshotAbi, spaces_ptr)];
#[cfg(target_pointer_width = "64")]
const _: [(); 8] = [(); std::mem::offset_of!(MwmDesktopSnapshotAbi, spaces_len)];
#[cfg(target_pointer_width = "64")]
const _: [(); 16] = [(); std::mem::offset_of!(MwmDesktopSnapshotAbi, windows_ptr)];
#[cfg(target_pointer_width = "64")]
const _: [(); 24] = [(); std::mem::offset_of!(MwmDesktopSnapshotAbi, windows_len)];
#[cfg(target_pointer_width = "64")]
const _: [(); 32] = [(); std::mem::offset_of!(MwmDesktopSnapshotAbi, focused_window_id)];

#[repr(C)]
#[derive(Debug, Clone, Default)]
pub struct MwmStatus {
    pub code: i32,
    pub message_ptr: *mut c_char,
}

impl MwmStatus {
    pub(crate) fn ok() -> Self {
        Self::default()
    }

    pub(crate) fn unavailable() -> Self {
        Self {
            code: MWM_STATUS_UNAVAILABLE,
            message_ptr: std::ptr::null_mut(),
        }
    }

    pub(crate) unsafe fn message(&self) -> Option<String> {
        (!self.message_ptr.is_null()).then(|| unsafe {
            CStr::from_ptr(self.message_ptr)
                .to_string_lossy()
                .into_owned()
        })
    }
}

#[repr(C)]
#[derive(Debug, Clone, Copy, Default)]
pub struct MwmRectAbi {
    pub x: i32,
    pub y: i32,
    pub width: i32,
    pub height: i32,
}

#[repr(C)]
#[derive(Debug, Clone, Copy, Default)]
pub struct MwmSpaceAbi {
    pub id: u64,
    pub display_index: usize,
    pub active: u8,
    pub kind: i32,
}

#[repr(C)]
#[derive(Debug, Clone, Default)]
pub struct MwmWindowAbi {
    pub id: u64,
    pub pid: u32,
    pub has_pid: u8,
    pub app_id_ptr: *mut c_char,
    pub title_ptr: *mut c_char,
    pub frame: MwmRectAbi,
    pub has_frame: u8,
    pub level: i32,
    pub space_id: u64,
    pub order_index: usize,
    pub has_order_index: u8,
}

#[repr(C)]
#[derive(Debug, Clone, Default)]
pub struct MwmDesktopSnapshotAbi {
    pub spaces_ptr: *mut MwmSpaceAbi,
    pub spaces_len: usize,
    pub windows_ptr: *mut MwmWindowAbi,
    pub windows_len: usize,
    pub focused_window_id: u64,
}

impl MwmDesktopSnapshotAbi {
    pub(crate) fn empty() -> Self {
        Self::default()
    }

    pub(crate) fn validate(&self) -> Result<(), &'static str> {
        if self.spaces_len > 0 && self.spaces_ptr.is_null() {
            return Err("spaces_ptr was null for a non-empty snapshot");
        }

        if self.windows_len > 0 && self.windows_ptr.is_null() {
            return Err("windows_ptr was null for a non-empty snapshot");
        }

        Ok(())
    }
}

impl MwmRectAbi {
    fn to_native_bounds(self) -> NativeBounds {
        NativeBounds {
            x: self.x,
            y: self.y,
            width: self.width,
            height: self.height,
        }
    }
}

impl MwmSpaceAbi {
    fn kind(self) -> SpaceKind {
        match self.kind {
            0 => SpaceKind::Desktop,
            1 => SpaceKind::Fullscreen,
            2 => SpaceKind::SplitView,
            3 => SpaceKind::System,
            4 => SpaceKind::StageManagerOpaque,
            _ => SpaceKind::System,
        }
    }
}

impl MwmDesktopSnapshotAbi {
    pub(crate) unsafe fn to_native_snapshot(&self) -> NativeDesktopSnapshot {
        let spaces = if self.spaces_len == 0 {
            &[][..]
        } else {
            unsafe { std::slice::from_raw_parts(self.spaces_ptr, self.spaces_len) }
        }
        .iter()
        .copied()
        .map(|space| NativeSpaceSnapshot {
            id: space.id,
            display_index: space.display_index,
            active: space.active != 0,
            kind: space.kind(),
        })
        .collect::<Vec<_>>();

        let windows = if self.windows_len == 0 {
            &[][..]
        } else {
            unsafe { std::slice::from_raw_parts(self.windows_ptr, self.windows_len) }
        }
        .iter()
        .map(|window| NativeWindowSnapshot {
            id: window.id,
            pid: (window.has_pid != 0).then_some(window.pid),
            app_id: unsafe { owned_string(window.app_id_ptr) },
            title: unsafe { owned_string(window.title_ptr) },
            bounds: (window.has_frame != 0).then_some(window.frame.to_native_bounds()),
            level: window.level,
            space_id: window.space_id,
            order_index: (window.has_order_index != 0).then_some(window.order_index),
        })
        .collect::<Vec<_>>();

        NativeDesktopSnapshot {
            active_space_ids: spaces
                .iter()
                .filter(|space| space.active)
                .map(|space| space.id)
                .collect(),
            spaces,
            windows,
            focused_window_id: (self.focused_window_id != 0).then_some(self.focused_window_id),
        }
    }
}

unsafe fn owned_string(ptr: *mut c_char) -> Option<String> {
    (!ptr.is_null()).then(|| unsafe { CStr::from_ptr(ptr).to_string_lossy().into_owned() })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ffi_transport_layout_matches_swift_contract() {
        assert_eq!(std::mem::size_of::<MwmStatus>(), 16);
        assert_eq!(std::mem::align_of::<MwmStatus>(), 8);
        assert_eq!(std::mem::offset_of!(MwmStatus, code), 0);
        assert_eq!(std::mem::offset_of!(MwmStatus, message_ptr), 8);

        assert_eq!(std::mem::size_of::<MwmRectAbi>(), 16);
        assert_eq!(std::mem::align_of::<MwmRectAbi>(), 4);
        assert_eq!(std::mem::offset_of!(MwmRectAbi, x), 0);
        assert_eq!(std::mem::offset_of!(MwmRectAbi, y), 4);
        assert_eq!(std::mem::offset_of!(MwmRectAbi, width), 8);
        assert_eq!(std::mem::offset_of!(MwmRectAbi, height), 12);

        assert_eq!(std::mem::size_of::<MwmSpaceAbi>(), 24);
        assert_eq!(std::mem::align_of::<MwmSpaceAbi>(), 8);
        assert_eq!(std::mem::offset_of!(MwmSpaceAbi, id), 0);
        assert_eq!(std::mem::offset_of!(MwmSpaceAbi, display_index), 8);
        assert_eq!(std::mem::offset_of!(MwmSpaceAbi, active), 16);
        assert_eq!(std::mem::offset_of!(MwmSpaceAbi, kind), 20);

        assert_eq!(std::mem::size_of::<MwmWindowAbi>(), 80);
        assert_eq!(std::mem::align_of::<MwmWindowAbi>(), 8);
        assert_eq!(std::mem::offset_of!(MwmWindowAbi, id), 0);
        assert_eq!(std::mem::offset_of!(MwmWindowAbi, pid), 8);
        assert_eq!(std::mem::offset_of!(MwmWindowAbi, has_pid), 12);
        assert_eq!(std::mem::offset_of!(MwmWindowAbi, app_id_ptr), 16);
        assert_eq!(std::mem::offset_of!(MwmWindowAbi, title_ptr), 24);
        assert_eq!(std::mem::offset_of!(MwmWindowAbi, frame), 32);
        assert_eq!(std::mem::offset_of!(MwmWindowAbi, has_frame), 48);
        assert_eq!(std::mem::offset_of!(MwmWindowAbi, level), 52);
        assert_eq!(std::mem::offset_of!(MwmWindowAbi, space_id), 56);
        assert_eq!(std::mem::offset_of!(MwmWindowAbi, order_index), 64);
        assert_eq!(std::mem::offset_of!(MwmWindowAbi, has_order_index), 72);

        assert_eq!(std::mem::size_of::<MwmDesktopSnapshotAbi>(), 40);
        assert_eq!(std::mem::align_of::<MwmDesktopSnapshotAbi>(), 8);
        assert_eq!(std::mem::offset_of!(MwmDesktopSnapshotAbi, spaces_ptr), 0);
        assert_eq!(std::mem::offset_of!(MwmDesktopSnapshotAbi, spaces_len), 8);
        assert_eq!(std::mem::offset_of!(MwmDesktopSnapshotAbi, windows_ptr), 16);
        assert_eq!(std::mem::offset_of!(MwmDesktopSnapshotAbi, windows_len), 24);
        assert_eq!(
            std::mem::offset_of!(MwmDesktopSnapshotAbi, focused_window_id),
            32
        );
    }
}
